### ISSUE TITLE
Wrong Colon rule

### ISSUE THREAD

**Issue by dineybomfim** [realm/SwiftLint#637]
On Swift Std Lib, the Colon rule is much more complex than the one implemented here in SwiftLint.

The SwiftLing RegEx matches only the last char, a spaced colon and then a char again.
That means it will match a trigger for any of the following:
"let abc : Void"
"Class Abc : Void"
"Dictionary<String : String>"

On the above cases, only the first one should be triggered.

Here is a piece of the Swift Std Lib, showing us how this pattern is well defined by Apple.

`public struct Range<Element : ForwardIndexType> : Equatable, CollectionType, CustomStringConvertible, CustomDebugStringConvertible {
...
public subscript (position: Element) -> Element { get }
public subscript (_: Element._DisabledRangeIndex) -> Element { get }
...
}`

Based on all the Swift Std Lib, we can infere that the clear rule for colon is:
• For Type declarations (class, struct, enum, protocol, extension) use the " : " (space, colon, space)
• For variables, constants and parameters (functions and closures) use the ": " (colon, space)

**Comment by sroebert** [comment]
Would love to see this fixed as well.

**Comment by masters3d** [comment]
Wrong here is very subjective. The only place where I think it is helpful to have spaces is where there are angle brackets involved. 

```swift
  struct Box<Base : Hashable> : _AnyHashableBox {  
                  ^           ^  
...              maybe       yes  
```

**Comment by dineybomfim** [comment]
Well, again an example from the Apple library (Swift 3 now):

```
public struct Dictionary<Key : Hashable, Value> : Collection, ExpressibleByDictionaryLiteral {
    public typealias Element = (key: Key, value: Value)
    public init(minimumCapacity: Int)
    public var startIndex: DictionaryIndex<Key, Value> { get }
```

This does not seem subjective to me, seems very clear: for type declarations use 2 spaces, for other levels use 1 space. Nothing subjective, very straight forward btw.

**Comment by masters3d** [comment]
if it was _wrong_ the compiler would reject it :)

**Comment by jpsim** [comment]
SwiftLint is very clear on not being based on any style guide other than GitHub's:

> A tool to enforce Swift style and conventions, loosely based on [GitHub's Swift Style Guide](https://github.com/github/swift-style-guide).

Also, the Swift standard library is written in a very unique style, not matched by most non-Apple projects.

---

SwiftLint explicitly doesn't _define_ its own style guide to avoid the bikeshedding involved in those kinds of, very subjective, decisions. Ideally, SwiftLint wouldn't diverge from GitHub's Swift styleguide at all, and any stylistic rules would be defined in the style guide first before being enforced by SwiftLint. Since that's pretty process-heavy, we haven't done that historically.

**Comment by dineybomfim** [comment]
Lots of talks, few definitions, the final answer is not clear to me.

Let's be crystal clear, this is something the core team wants to use as default or not?

**Comment by fikus** [comment]
Given the lack of consensus, might I suggest making the colon rule configurable, and specifically allowing the specification of different rules for type declarations and variable declarations?

**Comment by SDGGiesbrecht** [comment]
The examples in the [Swift API Design Guidelines](https://swift.org/documentation/api-design-guidelines/) are consistently preceded by a space, but the examples in [The Swift Programming Language](https://developer.apple.com/library/content/documentation/Swift/Conceptual/Swift_Programming_Language/) never have a preceding space. (Disclaimer: I ⌘F’ed through dozens of them, but I didn’t check every last one.)

In short, there is no unanimously recommended style.

**Comment by jpsim** [comment]
> Given the lack of consensus, might I suggest making the colon rule configurable, and specifically allowing the specification of different rules for type declarations and variable declarations?

PRs welcome!

**Comment by paulofaria** [comment]
any PRs on this?
